### ISSUE TITLE
doctrine \Doctrine\DBAL\Query\QueryBuilder getSQL crash

### ISSUE THREAD

**Issue by shengjie** [doctrine/dbal#2330]
Hello, I've a problem while using QueryBuilder.

Following code crashed and nginx served a 502 bad gateway.

```php
$qb = $sqlConnection->createQueryBuilder();
$qb->select('u')->from('User', 'u')->where('u.id = 1')->andWhere('u.email is not null');
echo $qb->getSQL();
```

Doctrine/DBAL: v2.4.5
PHP: 7.0.3

I've investigate a little, found crash happens in `(string) $this->sqlParts['where']` (\Doctrine\DBAL\Query\QueryBuilder:1022)
which calls:

`\Doctrine\DBAL\Query\Expression\CompositeExpression::__toString`

I've put following code
```php
/**
     * Retrieves the string representation of this composite expression.
     *
     * @return string
     */
    public function __toString()
    {
        if (count($this->parts) === 1) {
            return (string) $this->parts[0];
        }
        //MY TEST CODE
        echo $this->parts[0];
        echo $this->parts[1]==$this ? "true": "false";
        die();
        //END MY TEST CODE

        return '(' . implode(') ' . $this->type . ' (', $this->parts) . ')';
    }
```

result: `u.id = 1true`

so i think this is a circular reference problem.

Did i used the query builder wrong ? or is it fixed in a newer version ?

Thanks for any help :)

**Comment by Ocramius** [comment]
Can you please replicate the issue against the [`QueryBuilderTest`](https://github.com/doctrine/dbal/blob/61e1f860a304c3fb6961ecbec36927b5f20b71c6/tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php)?

**Comment by shengjie** [comment]
```
shengjie@MacBook-Pro-de-Shengjie:~/dbal$ git status
HEAD detached at v2.4.5
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	mysql.phpunit.xml
	phpunit

nothing added to commit but untracked files present (use "git add" to track)
```

```
shengjie@MacBook-Pro-de-Shengjie:~/dbal$ ./phpunit -c mysql.phpunit.xml tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php 
PHPUnit 3.7.24 by Sebastian Bergmann.

Configuration read from /Users/shengjie/dbal/mysql.phpunit.xml

......Segmentation fault: 11
```

**Comment by Ocramius** [comment]
Does the QueryBuilder test fail like that with the unmodified test suite?

**Comment by shengjie** [comment]
Yes, fresh clone:
What i have done:

```sh
git clone https://github.com/doctrine/dbal.git
cd dbal/
git checkout tags/v2.4.5
ln -s vendor/phpunit/phpunit/phpunit.php phpunit
cp phpunit.xml.dist mysql.phpunit.xml
vim mysql.phpunit.xml
./phpunit -c mysql.phpunit.xml tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php 
```

```xml
<?xml version="1.0" encoding="utf-8"?>
<!--
    Use this configuration file as a template to run the tests against any dbms.
    Procedure:
        1) Save a copy of this file with a name of your chosing. It doesn't matter
           where you place it as long as you know where it is.
           i.e. "mysqlconf.xml" (It needs the ending .xml).
        2) Edit the file and fill in your settings (database name, type, username, etc.)
           Just change the "value"s, not the names of the var elements.
        3) To run the tests against the database type the following from within the
           tests/ folder: phpunit -c <filename> ...
           Example: phpunit -c mysqlconf.xml
-->
<phpunit backupGlobals="false"
         backupStaticAttributes="false"
         colors="true"
         convertErrorsToExceptions="true"
         convertNoticesToExceptions="true"
         convertWarningsToExceptions="true"
         processIsolation="false"
         stopOnFailure="false"
         syntaxCheck="false"
         bootstrap="./tests/Doctrine/Tests/TestInit.php"
>
  <php>
    <!-- "Real" test database -->
    <!-- Uncomment, otherwise SQLite runs
    <var name="db_type" value="pdo_mysql"/>
    <var name="db_host" value="localhost" />
    <var name="db_username" value="root" />
    <var name="db_password" value="" />
    <var name="db_name" value="doctrine_tests" />
    <var name="db_port" value="3306"/>
    -->
    <!--<var name="db_event_subscribers" value="Doctrine\DBAL\Event\Listeners\OracleSessionInit">-->
    
    <!-- Database for temporary connections (i.e. to drop/create the main database) -->
    <var name="tmpdb_type" value="pdo_mysql"/>
    <var name="tmpdb_host" value="localhost" />
    <var name="tmpdb_username" value="root" />
    <var name="tmpdb_password" value="rootpassword" />
    <var name="tmpdb_name" value="doctrine_tests_tmp" />
    <var name="tmpdb_port" value="3306"/>
  </php>

    <testsuites>
        <testsuite name="Doctrine DBAL Test Suite">
            <directory>./tests/Doctrine/Tests/DBAL</directory>
        </testsuite>
    </testsuites>
</phpunit>
```

**Comment by Ocramius** [comment]
This was already fixed in `master` and `2.5`, although I can't find the exact reference. Closing as `duplicate`.

**Comment by arnaud-lb** [comment]
For reference, this bug affects PHP 7.x, and has been fixed by 75402954517d749df3705ef21d53c1e1709a6347.
